### DISCUSSION TITLE
Trying to Speed RefindPlus Up

### DISCUSSION THREAD

**Discussion by vicbyrd** [dakanji/RefindPlus#132] [General]
Hello,
I have been using rEFInd (0.13.2) for about a year now and lately it's taking about 20 seconds to load the menu. Please don't beat me up to hard for being a distro hopper as I have 5 linux distros along with my 2 Windows partitions. Understandable that refind/RP would take a while to scan. I enabled the **scanfor** option and selected only firmware and boot time got a little better but still over 10 seconds. Would that be considered "normal"?

I "installed" RefindPlus_x64-001303_AC\ with the same, long boot time, result. Running the debug efi produces this log:
[https://1drv.ms/u/s!Ak62NP22_qybipxgtyCOdRqQoNjFcQ?e=wlCcVb](url)
again with only the firmware being selected in **scanfor**.

Questions:

1. Is ** Could Not Calibrate ACPI PM Timer ... Intel ICH Device Not Found ** a problem - I'm running an AMD processor?
2. Why the stall, per timestamp, between load drivers and Link Device Handles?
 ```
   0:737    0:023  L O A D   U E F I   D R I V E R S   :::::   P R O G R A M   D E F A U L T   F O L D E R
   0:755    0:017  Scan 'EFI\REFIND\drivers' Folder:
                     - Not Found or Empty
   0:764    0:009  Scan 'EFI\REFIND\drivers_x64' Folder:
                     - Success ... UEFI Driver:- 'EFI\REFIND\drivers_x64\ext4_x64.efi'
                     - Success ... UEFI Driver:- 'EFI\REFIND\drivers_x64\btrfs_x64.efi'

   0:811    0:046  INFO: Clear Keystroke Buffer ... Already Started
   8:974    8:162  L I N K   D E V I C E   H A N D L E S
```
3. Anyway to identify device @ Handle location 0x0F8 - seems to stall for a moment here?
```
  17:235    8:199  Handle 0x0F8 ... Bypassed [Not Linkable] - Not PCIe Device : Leverages GOP
  17:242    0:006  Handle 0x172 ... Bypassed [Not Linkable] - Not PCIe Device
```

Thanks for any info/insight that you might provide,
Victor

**Comment by dakanji** [comment]
Link is broken ... Why not just attach the log file instead?

**Comment by vicbyrd** [comment]
Here's the log:

[22x14r5653.log](https://github.com/dakanji/RefindPlus/files/10236862/22x14r5653.log)

Dakanji Edit: You can just drag and drop files to attach them

**Comment by dakanji** [comment]
**Your Query 1:**
The timer thing affects the accuracy of the timer and timestamps ... Not a serious issue

**Your Queries 2 & 3:**
The apparent stall between loading drivers and the `LINK DEVICE HANDLES` entry on one hand and that on `Handle 0x0F8` on the other, are one and the same; which is that there is a delay in processing that handle.

The connection to the former is that the Device Handle connection is done twice but the first loop through is not logged.

So, it went through the first "Not Logged" loop, hit the delay on that handle (reflected when the logging started) and then hit the delay the second time. Basically, that handle is responsible for about 15/16 secs of the 20/21 sec loading time.

BTW, the reason for a double loop is that some handles are only revealed after some device handles are linked.

As for identifying what is at Handle 0x0F8, you will probably need someone like @joevt and one of his magic EFI Shell scripts.

**Reply by joevt** [reply]
> As for identifying what is at Handle 0x0F8, you will probably need someone like @joevt and one of his magic EFI Shell scripts.

When you get into RefindPlus, you can go into the UEFI Shell and use the `dh -d -v > dh_d_v.txt` command to output all the handles to a text file. But first you need to select the EFI partition as the current file system. Type `ls fs0:` (or maybe `ls fs0:/`) and repeat with different numbers for the file system until you find the EFI partition.
 https://sourceforge.net/p/refind/discussion/general/thread/f1f144d655/

**Reply by vicbyrd** [reply]
Thanks for the process. Gave it a shot but shell would not let me redirect the dh -d -v command to text file. Gave me an error. I did try **dh F8** for "detailed" information re that handle address which gave me the output:
    F8: DevicePath (PciRoot(0x0)   PCIRootBridgeIO (PCIRootBridgeIO)
Don't know that I can do anything with that since it's not a device I can disable (e.g. pull out a bad usb drive). Maybe try a bios update since I am running behind one update - which I do as a common practice.
Thanks in advance for your assistance.

**Reply by joevt** [reply]
You must find a file system that is writable to get redirect to work. Usually only FAT file systems are writable. This usually includes EFI partitions.
```
`fs0:`
`ls`
dh -d -v > dh_d_v.txt
```
`fs0` might not be an EFI partition, so you have to try the other file system numbers.
`fs1:`
`fs2:`
etc.

Some EFI shells might use hex numbers, so after `fs9` comes `fsA` to `fsF`, then `fs10`. Check the `map` command for a list of file system numbers.

**Comment by dakanji** [comment]
I forgot to mention that you can force a single Device Handle connection loop by activating `disable_rescan_dxe`.
You can read up on this setting in the [sample config file](https://github.com/dakanji/RefindPlus/blob/5f55917281144f9c0a85d172fc77071a6e4e3391/config.conf-sample#L205-L218)

**Comment by vicbyrd** [comment]
Thanks! I’ll un-comment that line and post results.

**Comment by vicbyrd** [comment]
New log follows. disable_rscan_dxe cut boot time from 20 to 11 seconds. Nice! Still stalling @ 
    8:658    8:163  Handle 0x0F8 ... Bypassed [Not Linkable] - Not PCIe Device : Leverages GOP
will try to contact @joevt for help - thanks for the referral.

[22x15m1226.log](https://github.com/dakanji/RefindPlus/files/10239136/22x15m1226.log)

DA Edit: Recreated and attached log

**Comment by dakanji** [comment]
Pasting the contents of an entire log into a post, here or elsewhere, is essentially an assault on other readers.
Please always attach log files instead.

Amended your post(s) accordingly

**Comment by vicbyrd** [comment]
Sorry 'bout that - tried to drag and drop but got an error. Will be aware of that in the future.
BTW thanks for the help!

**Comment by vicbyrd** [comment]
Original issue = 8:747    8:177  Handle 0x0F8 ... Bypassed [Not Linkable] - Not PCIe Device : Leverages GOP (from attached log)

fs0: was the correct file system – I was trying to run the dh command from the ‘SHELL’ prompt instead of changing to the fs0: prompt.
Command output is for handle 0x0F8 (subset of attached .txt) is:

**DA Edit: Reconstituted dump into a file and attached this instead:**
[00_Dump1.txt](https://github.com/dakanji/RefindPlus/files/10255938/00_Dump1.txt)

Obviously enumerating the devices hanging off the PCIe bus. Should this process take 8 seconds?
The total boot time is 14 seconds would love to find a way to shave this 8 seconds but I don’t know if that’s possible.
I’ve updated my bios to the latest available (AMD_X570-p) don’t know what else I can do.
I’ve attached my latest boot log in case you can see anything else I can try.

Thanks for helping me out,
Victor

**DA Edit: Removed Email Junk**

﻿Handle Dump
**DA Edit: Reconstituted dump into a file and attached this instead:**
[00_Dump2.txt](https://github.com/dakanji/RefindPlus/files/10255904/00_Dump2.txt)

**Reply by joevt** [reply]
Don't use e-mail for replying to issues.
Do use attachments for long log output. Your log was cutoff at handle 92.
If handle f8 is not the problem, then it could be the one that's processed before that which is handle e8 in the [22x15m1226.log](https://github.com/dakanji/RefindPlus/files/10239136/22x15m1226.log) that you posted. It depends on the order of the logging messages which I haven't looked at recently. I remember changing some logging in this area in my fork of RefindPlus but I don't know if those changes would be more informative here.
